feature: localcache

codus-me opened the pull request:
宣告 localcache interface
實現 localcache
測試 localcache

codus-me committed:
feature: localcache

VineyAT commented on go.mod:
```diff
@@ -0,0 +1,11 @@
+module codus.me/localcache
+
+go 1.19
```
這個版本是最新的，不過未來我們開發會選用 1.18，嘗試先 downgrade 一下

codus-me replied:
收到，已安裝 go1.18

VineyAT commented on localcache_test.go:
```diff
@@ -0,0 +1,37 @@
+package localcache
```
通常呢，test case 會跟隨實作，也就是你真正要寫 test case 的是 implement_test.go 而不是 localcache.go，因為他只是一個 interface

codus-me replied:
收到，感謝。確實是，會再修正

VineyAT commented on localcache_test.go:
```diff
@@ -0,0 +1,37 @@
+}
+
+func (suite *LocalcacheTestSuite) SetupTest() {
+	suite.VariableThatShouldStartAtFive = 5
```
What's this? do you use it?

codus-me replied:
拍寫，這是從 testify/suite 複製過來，漏刪了。會再修正

VineyAT commented on implement.go:
```diff
@@ -0,0 +1,30 @@
+
+func (obj *cacheImpl) Set(key string, value interface{}) {
+	now := time.Now()
+	obj.createdAt[key] = &now
```
為什麼同一把 Key 要用兩個 map 存？這樣就存在 racing issue，建議合再一起

codus-me replied:
收到，感謝。確實是，會再修正

VineyAT commented on localcache.go:
```diff
@@ -0,0 +1,20 @@
+}
+
+// TTL time to live 30 seconds
+const TTL time.Duration = 30 * time.Second
```
基本上他不用曝露出來，你可以預設一個 default，在你 New 的當下給對方帶進來，如果給空值就給 default 是一個設計

codus-me replied:
收到，原本是想說放外面揭露資訊讓呼叫方知道 TTL 即可。

VineyAT commented on localcache_test.go:
```diff
@@ -0,0 +1,37 @@
+	suite.Equal(nil, suite.cache.Get("not exist"))
+}
+func (suite *LocalcacheTestSuite) TestLocalcacheSetThenGet() {
+	suite.cache.Set("mykey", 1)
```
這邊有點矛盾，你用 Get 去驗證 Set? 但是並不能代表哪一個是無疑慮的

以你的結構而言，應該是塞進去到 map 在用 Get() 驗證？或是反過來，Set 塞進去再用讀取 map 的方式驗證

codus-me replied:
收到，確實是，會再修正

VineyAT commented on localcache_test.go:
```diff
@@ -0,0 +1,37 @@
+	"github.com/stretchr/testify/suite"
+)
+
+type LocalcacheTestSuite struct {
```
小寫即可

codus-me replied:
收到，感謝，確實是，會再修正

VineyAT commented on localcache.go:
```diff
@@ -0,0 +1,20 @@
+const TTL time.Duration = 30 * time.Second
+
+// New will create and return a implementation of Cache
+func New() Cache {
```
New 搬進去 impl 裡面，他不屬於 Interface

codus-me replied:
收到，感謝。確實是，會再修正

VineyAT commented on localcache.go:
```diff
@@ -0,0 +1,20 @@
+
+// New will create and return a implementation of Cache
+func New() Cache {
+	return newCacheImpl()
```
其實我不是很懂，何不把 newCacheImpl() 內容直接搬進來即可，然後把 New() 放到 impl.go 裡面

codus-me replied:
收到，感謝。確實是，會再修正

VineyAT commented on localcache_test.go:
```diff
@@ -0,0 +1,37 @@
+	suite.cache.Set("mykey", 1)
+	impl := suite.cache.(*cacheImpl)
+	impl.createdAt["mykey"] = &time.Time{}
+	suite.Equal(nil, suite.cache.Get("mykey"))
```
改成用 `s.Require().Equal()` 會好一點，這樣一有錯誤就會停下來，不會繼續跑下去

codus-me replied:
收到，感謝

codus-me committed:
feature: downgrade go to v1.18

codus-me committed:
refactor: use lower case and remove useless code

codus-me committed:
optimize: use suite.Require().Equal instead suite.Equal

codus-me committed:
feature: move ttl to implementation (and private it)
fix: testcase set not use get to validate, access hashmap value directly instead
refactor: move New into impl
refactor: cache save struct including data and createdAt instead seperrate

codus-me committed:
fix: test implementation instead interface

codus-me committed:
test: add testcase get

codus-me committed:
refactor: change testcase order. move set blow get

codus-me committed:
refactor: use variable with meaningful name instead code in if condition

VineyAT commented on implement.go:
```diff
@@ -0,0 +1,41 @@
+	if obj.hashMap[key] == nil {
+		return nil
+	}
+	outdated := time.Now().After(obj.hashMap[key].createdAt.Add(ttl))
```
這邊拆成兩個動作，一個拿 createdAt 另一個拿 data，或許可以一次取值，確保都是時刻的資料

VineyAT replied:
另外你的 cache 是沒辦法確認成 thread safe 的，多個請求背後使用同一個 cache 資料會不一致，建議加個 lock，你可以查詢一下

codus-me replied:
確實是，我再拆成一次取值，感謝。
thread safe 其實有想到，但我想說先把code寫好再來加，感謝提醒><

VineyAT commented on implement.go:
```diff
@@ -0,0 +1,41 @@
+)
+
+// New will create and return a implementation of Cache
+func New() *cacheImpl {
```
func New() Cache { ... }

cacheImpl 是小寫，沒辦法給外部使用，直接改成 Cache 也就是 interface

codus-me replied:
感謝提醒，會再修正

VineyAT reviewed (approved):
ac with comments

codus-me committed:
fix: New return interface instead private cacheImpl

codus-me committed:
refactor: cacheImpl get data one time and saved to variable

codus-me committed:
feature: let cache thread safe through RWLock

VineyAT commented on implement.go:
```diff
@@ -0,0 +1,48 @@
+func (obj *cacheImpl) Get(key string) interface{} {
+	obj.mux.RLock()
+	defer obj.mux.RUnlock()
+	cachedData := obj.hashMap[key]
```
```go
cachedData, ok := obj.hashMap[key]
if !ok {
	return nil
} else if cachedData == nil {
	return nil
}
```

VineyAT commented on implement.go:
```diff
@@ -0,0 +1,48 @@
+	if cachedData == nil {
+		return nil
+	}
+	outdated := time.Now().After(cachedData.createdAt.Add(ttl))
```
```go
if ok := time.Now().After(cachedData.createdAt.Add(ttl)); !ok {
	return nil
}
```

VineyAT reviewed (approved):
ac with comment

VineyAT commented on implement_test.go:
```diff
@@ -0,0 +1,59 @@
+	var wg sync.WaitGroup
+	wg.Add(200)
+	for i := 0; i < 100; i++ {
+		go func() {
```
wg.Add(1)

VineyAT commented on implement_test.go:
```diff
@@ -0,0 +1,59 @@
+}
+func (suite *localcacheTestSuite) TestLocalcacheConcurrency() {
+	var wg sync.WaitGroup
+	wg.Add(200)
```
remove this

codus-me committed:
refactor: use channel to test concurrency set

codus-me committed:
refactor: use if ok = ... to validate data

codus-me committed:
feature: add Fetch

codus-me committed:
feature: golang 1.18 -> 1.22.6
feature: add Do (singleflight)
